File: app/src/main/java/dev/suhockii/lifetest/ui/products/ProductsPresenter.java
package dev.suhockii.lifetest.ui.products;

import android.view.View;

import moxy.InjectViewState;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import dev.suhockii.lifetest.model.Product;
import dev.suhockii.lifetest.repo.AppRepository;
import dev.suhockii.lifetest.util.mvp.NonLeakPresenter;
import dev.suhockii.lifetest.util.rx_transformers.RxSchedulers;
import io.reactivex.Single;
import timber.log.Timber;

import static dev.suhockii.lifetest.di.module.RepositoryModule.LOCAL_REPOSITORY;
import static dev.suhockii.lifetest.di.module.RepositoryModule.REMOTE_REPOSITORY;

@InjectViewState
public class ProductsPresenter extends NonLeakPresenter<ProductsView> {

    private final AppRepository localRepository;
    private final AppRepository remoteRepository;
    private final RxSchedulers rxSchedulers;

    @Inject
    public ProductsPresenter(@Named(LOCAL_REPOSITORY) AppRepository localRepository,
                             @Named(REMOTE_REPOSITORY) AppRepository remoteRepository,
                             RxSchedulers rxSchedulers) {
        this.localRepository = localRepository;
        this.remoteRepository = remoteRepository;
        this.rxSchedulers = rxSchedulers;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        loadProducts();
    }

    private void loadProducts() {
        getViewState().showProgressBar(View.VISIBLE);

        localRepository.getProducts()
                .onErrorResumeNext(localEmpty -> updateProducts())
                .compose(rxSchedulers.getIoToMainTransformerSingle())
                .doOnSubscribe(compositeDisposable::add)
                .doOnEvent((products, throwable) -> getViewState().showProgressBar(View.INVISIBLE))
                .doOnSuccess(products -> getViewState().clearShowSnackbarCommand())
                .doOnError(throwable -> getViewState().showSnackbar(throwable.getMessage(), this::loadProducts))
                .subscribe(products -> getViewState().showProducts(products), Timber::e);
    }

    private Single<List<Product>> updateProducts() {
        return remoteRepository.getProducts()
                .doOnSuccess(localRepository::saveProducts);
    }
}

File: app/src/main/java/dev/suhockii/lifetest/ui/products/ProductsPresenter.kt
package dev.suhockii.lifetest.ui.products

import android.view.View

import moxy.InjectViewState

import javax.inject.Inject
import javax.inject.Named

import dev.suhockii.lifetest.model.Product
import dev.suhockii.lifetest.repo.AppRepository
import dev.suhockii.lifetest.util.mvp.NonLeakPresenter
import dev.suhockii.lifetest.util.rx_transformers.RxSchedulers
import io.reactivex.Single
import timber.log.Timber

import dev.suhockii.lifetest.di.module.RepositoryModule.LOCAL_REPOSITORY
import dev.suhockii.lifetest.di.module.RepositoryModule.REMOTE_REPOSITORY
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer

@InjectViewState
class ProductsPresenter @Inject constructor(
    @param:Named(LOCAL_REPOSITORY) private val localRepository: AppRepository,
    @param:Named(REMOTE_REPOSITORY) private val remoteRepository: AppRepository,
    private val rxSchedulers: RxSchedulers
) : NonLeakPresenter<ProductsView>() {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadProducts()
    }

    private fun loadProducts() {
        viewState.showProgressBar(View.VISIBLE)

        localRepository.products
            .onErrorResumeNext { updateProducts() }
            .compose(rxSchedulers.getIoToMainTransformerSingle())
            .doOnSubscribe { compositeDisposable.add(it) }
            .doOnEvent { _, _ -> viewState.showProgressBar(View.INVISIBLE) }
            .doOnSuccess { viewState.clearShowSnackbarCommand() }
            .doOnError { throwable ->
                Timber.e(throwable)
                viewState.showSnackbar(throwable.message, Runnable { this.loadProducts() })
            }
            .subscribe { products -> viewState.showProducts(products) }
    }

    private fun updateProducts(): Single<List<Product>> {
        return remoteRepository.products
            .doOnSuccess { localRepository.saveProducts(it) }
    }
}
